Framework of passive-data collectors

peculiar opened the pull request:
Description:
Providing a generic framework/collection of passive-data collectors of which there are specific subclasses to collect HealthKit, displacement, and motion activity data.

The collectors have a data sink (some share a data sink) that conforms to a protocol that is used to handle the collected data. This divorces collectors from the handling of the collected data thus providing flexibility.

The HealthKit, displacement, and Motion Activity collectors derive from the generic data collector. Both the HealthKit and Motion Activity collector use the generic data sink to handle writing data to a csv file. The data sink can be derived from to further enhance/extend the data handling to do things like hold data types in memory, write to core data, etc...

The Displacement collector uses a sink derived from the generic data sink in order to customize the transformation of CLLocation data to a stream of displacement vectors, during the life of the collector.
How it was tested:

HealthKit Collector - This was tested with a real device over more than 24 hours, manually and passively inputting data. Observing data as it is written to file.

Displacement Collector - Tested in a vehicle, walking, resting and observing data as it is written to file.
CoreMotionActivity Collector - Observing data a- s it is written to file.

Notes:
- Background entitlements are necessary.
- Displacement collector requires alwaysInUse permissions.
- If HealthKit permissions to types are not given then no data is collected. 
- HealthKit Collector only supports Category, Quantity, and HKWorkout types.

Known issues:
- The Displacement collector uses a sink derived from the generic data sink in order to customize the transformation of CLLocation data but this should be refactored since the derivation doesn't extend the functionality of the sink, rather it extends the functionality of a block that takes CLLocation data and transforms it into vector information.

peculiar committed:
Merge pull request #1 from ResearchKit/master

Update repo from ORK repo

peculiar committed:
Passive data-collection framework, HealthKit and location

commit 21292fe997d096842f1306ce7d8289796ffd9dae
Author: Edward Cessna <edward.cessna@ymedialabs.com>
Date:   Wed May 13 17:12:42 2015 -0700

    Passive data-collection framework

peculiar committed:
Passive data-collection framework, HealthKit and location

commit 21292fe997d096842f1306ce7d8289796ffd9dae
Author: Edward Cessna <edward.cessna@ymedialabs.com>
Date:   Wed May 13 17:12:42 2015 -0700

    Passive data-collection framework

peculiar committed:
Merge branch 'master' of https://github.com/peculiar/AppCore

# Conflicts:
#	APCAppCore/APCAppCore/UI/Onboarding/SignUp/APCAllSetContentViewController.m

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/CLLocation+APCBearing.m:
```diff
@@ -0,0 +1,80 @@
+    double lonDeltaSquared  = pow(lon2 - lon1, 2);
+    double latLonSummed     = latDeltaSquared + lonDeltaSquared;
+    
+    return sqrt(latLonSummed);
```
I don't see this method being used anywhere, and it's problematic. You're using L2 distance metrics, in a space that wraps. Best removed?

peculiar replied:
Will remove.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/HKWorkout+APCHelper.m:
```diff
@@ -0,0 +1,111 @@
+
+@implementation HKWorkout (APCHelper)
+
++ (NSString*)workoutActivityTypeStringRepresentation:(int)num
```
A framework defining categories on another framework's classes ought to use a prefix on method names, to avoid collisions. e.g., apc_workoutActivityType....

peculiar replied:
Will add "apc_" prefix.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NSDictionary+APCStringify.m:
```diff
@@ -0,0 +1,62 @@
+
+@implementation NSDictionary (APCStringify)
+
++ (NSString*) convertDictionary:(NSDictionary *)dict ToStringWithReturningError: (NSError * __autoreleasing *) errorToReturn
```
Also this isn't a canonical method name: e.g., capitalization of "To"; something like `stringFromDictionary:error:` might be better?

peculiar replied:
Absolutely agree!

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NSDictionary+APCStringify.m:
```diff
@@ -0,0 +1,62 @@
+                                                      encoding:NSUTF8StringEncoding];
+    }
+    
+    return !errorToReturn ? stringRepresentation : @"";
```
This is fragile. Surely stringRepresentation should not be created from jsonData at all if there was an error.

peculiar replied:
I will remove this functionality as it's no longer needed.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/APCPassiveDataCollector.h:
```diff
@@ -1,44 +1,45 @@
+@interface APCPassiveDataCollector : NSObject
 
-@property (nonatomic, strong) HKUnit* unitForTracker;
+- (void)addDataSink:(id)dataSync;
```
Untyped? really?

peculiar replied:
Will specify type as `APCDataCollector*`

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/APCPassiveDataCollector.m:
```diff
@@ -0,0 +1,81 @@
+{
+    for (APCDataCollector* collector in self.dataSinkList)
+    {
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^
```
This, and the other similar logic in startCollecting, is dispatching separate blocks onto a background queue for each collector. Why not make a copy of the data sink list, and dispatch just once?

peculiar replied:
Will invert the processing.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/APCPassiveDisplacementTrackingDataUploader.m:
```diff
@@ -0,0 +1,293 @@
+@implementation APCPassiveDisplacementTrackingDataUploader
+@synthesize baseTrackingLocation = _baseTrackingLocation;
+@synthesize mostRecentUpdatedLocation = _mostRecentUpdatedLocation;
+- (NSArray*)locationDictionaryWithLocationManager:(CLLocationManager*)manager
```
This returns an array, not a dictionary

peculiar replied:
Will rename.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/APCPassiveDisplacementTrackingDataUploader.m:
```diff
@@ -0,0 +1,293 @@
+        __typeof(self)  strongSelf  = weakSelf;
+        NSArray*        result      = nil;
+        
+        if ((self.baseTrackingLocation.coordinate.latitude == 0.0) && (self.baseTrackingLocation.coordinate.longitude == 0.0))
```
Why the comparisons with 0.0? Shouldn't baseTrackingLocation just be nil if invalid?

peculiar replied:
Good catch. Will change the if statement to check baseTrackingLocation property for nil value.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/APCPassiveDisplacementTrackingDataUploader.m:
```diff
@@ -0,0 +1,293 @@
+        //Send to delegate
+        if (result)
+        {  
+            NSString *stringToWrite = [NSString stringWithFormat:@"%@,%@,%@,%@,%@,%@,%@,%@,%@,%@,%@,%@,%@,%@\n",
```
This looks fragile! Why not create a mutable string and append each element in a loop?

peculiar replied:
Will replace using componentsJoinedByString method.

jwe-apple replied:
I specifically didn't say that because you'll also need the new-line and it makes no sense to copy the entire string just to do that.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/APCPassiveDisplacementTrackingDataUploader.m:
```diff
@@ -0,0 +1,293 @@
+            self.baseTrackingLocation = self.mostRecentUpdatedLocation;
+            self.mostRecentUpdatedLocation = manager.location;
+            
+            CLLocationDistance  distanceFromReferencePoint = [self.baseTrackingLocation distanceFromLocation:manager.location];
```
This code is effectively the same in both cases and should be pulled out.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/APCPassiveDisplacementTrackingDataUploader.m:
```diff
@@ -0,0 +1,293 @@
+                    NSDictionary* dict      = nil;
+
+                    dict                        = [NSDictionary dictionaryWithJSONString:jsonString];
+                    _mostRecentUpdatedLocation  = [[CLLocation alloc] initWithLatitude:[dict[kLat] doubleValue]
```
This is an almost an exact copy of the code for baseTrackingLocation, with a different path. Needs factoring out into a single method.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/APCPassiveDisplacementTrackingDataUploader.m:
```diff
@@ -0,0 +1,293 @@
+    
+    [self.healthKitCollectorQueue addOperationWithBlock:^
+    {
+        __typeof(self)  strongSelf  = weakSelf;
```
You do this, and then access self directly. I think accessing self directly should be fine, so why are you doing this?

peculiar replied:
I believe the intent is to keep the uploader object alive (retain count > 0)  while the block within the queue are active and there are no other references to the uploader.

jwe-apple replied:
Merely accessing self would accomplish that.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/APCPassiveDisplacementTrackingDataUploader.m:
```diff
@@ -0,0 +1,293 @@
+{
+    __weak typeof(self) weakSelf = self;
+    
+    [self.healthKitCollectorQueue addOperationWithBlock:^
```
This queue is doing otherwise unprotected operations on self. Is it backed by a serial queue?

peculiar replied:
It is a sequential queue.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCCoreMotionBackgroundDataCollector.m:
```diff
@@ -0,0 +1,153 @@
+        
+        __weak typeof(self) weakSelf = self;
+        
+        [self.motionActivityManager queryActivityStartingFromDate:lastTrackedEndDate
```
What ensures that these updates arrive and get processed before the activity updates start getting processed? I think this matters for activity data, because the activity objects are only valid relative to the previous/next object.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCDataCollector.h:
```diff
@@ -0,0 +1,57 @@
+#import <Foundation/Foundation.h>
+#import "APCCollectorProtocol.h"
+
+typedef NSDate*(^InitialStartDatePredicateDesignator)();
```
Since this is framework code, needs a prefix.

peculiar replied:
Adding "APC" prefix.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCHealthKitBackgroundDataCollector.m:
```diff
@@ -0,0 +1,208 @@
+    if (self)
+    {
+        _sampleType         = type;
+        _healthStore        = [HKHealthStore new];
```
I think it would be better if you used a common healthstore across these collectors, i.e. create it outside and pass it into each at init time.

peculiar replied:
Agree: this would be a better approach. Will create an issue to track this.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCHealthKitBackgroundDataCollector.m:
```diff
@@ -0,0 +1,208 @@
+    
+    self.observerQuery = [[HKObserverQuery alloc] initWithSampleType:sampleType
+                                                                 predicate:nil
+                                                             updateHandler:^(HKObserverQuery __unused *query,
```
Why is the query marked unused? It looks like it is used.

peculiar replied:
Removed the modifier.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCHealthKitBackgroundDataCollector.m:
```diff
@@ -0,0 +1,208 @@
+
+        if (sampleKind)
+        {
+            if ([sampleKind isKindOfClass:[HKCategorySample class]])
```
All this code is here purely for generating console output?

peculiar replied:
Will remove logging code.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCHealthKitBackgroundDataCollector.m:
```diff
@@ -0,0 +1,208 @@
+}
+
+
+- (void)anchorQuery:(HKObserverQuery*)query completionHandle:(HKObserverQueryCompletionHandler)completionHandler
```
Method name should be completionHandler:

peculiar replied:
Corrected typo.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCHealthKitDataBridge.h:
```diff
@@ -1,47 +1,52 @@
--(void) addTracker: (APCDataTracker*) tracker;
 
-- (void) flush:(APCDataTracker*) tracker; //Packages the data collected so far as encrypted zip.
+- (instancetype) initWithIdentifier: (NSString *) identifier andSampleType: (HKSampleType *) sampleType;
```
The 'and' in the method names here is not canonical

peculiar replied:
The file is cruft and is not part of the project. Deleting the header and corresponding implementation file.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Data Sink/APCPassiveDataSink.m:
```diff
@@ -0,0 +1,472 @@
+        
+        if (![[NSFileManager defaultManager] createDirectoryAtPath:path
+                                       withIntermediateDirectories:YES
+                                                        attributes:@{ NSFileProtectionKey : NSFileProtectionCompleteUntilFirstUserAuthentication }
```
Unconvinced that this is sufficient file protection for health data. Shouldn't protection be Complete, or CompleteUnlessOpen?

Of course once you do that there is the chance, at any time, that you might not be able to read the data if operating in the background.

peculiar replied:
Background operation is why protect is set to NSFileProtectionCompleteUntilFirstUserAuthentication.

DanKeen replied:
Health data should be stored at `NSFileProtectionCompleteUnlessOpen` or higher. You can use `UnlessOpen` to journal data prior to an unlock securely and then process it at the next unlock.

peculiar replied:
Received and understood. Will create an ticket to track resolution of this issue.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Data Sink/APCPassiveDataSink.m:
```diff
@@ -0,0 +1,472 @@
+                                       withIntermediateDirectories:YES
+                                                        attributes:@{
+                                                                     NSFileProtectionKey :
+                                                                         NSFileProtectionCompleteUntilFirstUserAuthentication
```
Same file protection comment applies here, although not obvious that you want to be applying your file protection settings at the directory level.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Data Sink/APCPassiveDataSink.m:
```diff
@@ -0,0 +1,472 @@
+{
+    __weak typeof(self) weakSelf = self;
+    
+    [self.healthKitCollectorQueue addOperationWithBlock:^{
```
Seems really rather inefficient to do one block and one complete file-open / file-write / file-close, on a per-sample basis.

peculiar replied:
The parameter is poorly named: potentially an array of samples may be processed by this method.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Data Sink/APCPassiveDataSink.m:
```diff
@@ -0,0 +1,472 @@
+                                                 andTaskRunUuid:nil
+                                                 returningError:&error];
+    
+    if (!success)
```
Is this error really always unrecoverable?

peculiar replied:
This is a potential issue. Will raise an issue to track.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCHealthKitDataBridge.m:
```diff
@@ -0,0 +1,99 @@
+    
+    if (results)
+    {
+        [[NSNotificationCenter defaultCenter] postNotificationName:APCHealthKitObserverQueryUpdateForSampleTypeNotification
```
Why is this needed when you have a delegate receiving the data?

peculiar replied:
Left over code that's not part of the project but wasn't removed from the repository. Will be removed.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCHealthKitDataBridge.h:
```diff
@@ -1,47 +1,52 @@
+
+#import "APCCollectorProtocol.h"
+
+@interface APCHealthKitDataBridge : NSObject
```
Maybe I'm missing something - what instantiates this? Is it even used?

peculiar replied:
Left over code that's not part of the project but wasn't removed from the repository. Will be removed.

saniul commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Data Sink/APCCollectorProtocol.h:
```diff
@@ -0,0 +1,45 @@
+
+- (void)didRecieveUpdatedValuesFromCollector:(id)results;
+- (void)didRecieveUpdatedValueFromCollector:(id)result;
+- (void)didRecieveUpdateWithLocationManager:(CLLocationManager*)manager withUpdateLocations:(NSArray*)locations;
```
Typos in the above method names: s/recieve/receive

jwe-apple commented:
@peculiar Please let me know when you feel you're ready for this to be merged.

peculiar committed:
Address comments from PR

peculiar commented:
r=me
@jwe-apple - ready for merge.

Commit:
Addressed PR comments.

Correclty re-implemented NSDictionary category to transform dictionary into a string. Refactored displacement uploader. Passing in HealthStore as a parameter into Healthkit Collector. Refactored file security key so user can specificy value. And if archiver and uploader fails we are careful not to delete the data.

Commit:
Further refactoring of Displacement uploader.

Commit:
Addressing PR comments. Code formatting.

Commit:
Addressing PR comments.

peculiar committed:
Merge pull request #2 from jwarmkessel/master

Addressing PR comments.

peculiar commented on APCAppCore/APCAppCore/Library/Categories/CLLocation+APCBearing.m:
```diff
@@ -0,0 +1,67 @@
+static double DegreesToRadians(double degrees) {return degrees * M_PI / 180;};
+static double RadiansToDegrees(double radians) {return radians * 180/M_PI;};
+
+@implementation CLLocation (APCAdditions)
```
@jwarmkessel Is the category name correct? The file name says "APCBearing".

peculiar commented on APCAppCore/APCAppCore/Library/Categories/CLLocation+APCBearing.h:
```diff
@@ -1,42 +1,41 @@
+#import <CoreLocation/CoreLocation.h>
 
-@property (nonatomic, strong) CMMotionActivityManager *motionActivityManager;
+@interface CLLocation (APCAdditions)
```
@jwarmkessel Is the category name correct? The file name says "APCBearing".

peculiar commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCDisplacementTrackingCollector.m:
```diff
@@ -0,0 +1,175 @@
+// 
+//  APCCoreLocationTracker.m 
```
@jwarmkessel Comment does not match file name.

peculiar commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCDisplacementTrackingCollector.m:
```diff
@@ -0,0 +1,175 @@
+
+- (void)start
+{
+    if ([CLLocationManager locationServicesEnabled] == YES)
```
@jwarmkessel The comparison is not safe on 32-bit platforms (iPhone 4s).

peculiar commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCDisplacementTrackingCollector.m:
```diff
@@ -0,0 +1,175 @@
+
+- (void)stop
+{
+    if ([CLLocationManager locationServicesEnabled] == YES)
```
@jwarmkessel The comparison is not safe on 32-bit platforms (iPhone 4s). The same comment applies to the next if statement.

peculiar commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCHealthKitBackgroundDataCollector.h:
```diff
@@ -1,49 +1,49 @@
-// Copyright (c) 2015, Apple Inc. All rights reserved. 
-// 
+//
+//  APCHealthKitCumulativeQuantityTypeDataBridge.h
```
The comment does not match the file title.

peculiar commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Data Sink/APCPassiveDataSink.h:
```diff
@@ -0,0 +1,72 @@
+#import <Foundation/Foundation.h>
+#import "APCCollectorProtocol.h"
+
+typedef NSString* (^CSVSerializer)(id dataSample);
```
Should the typedef have the "APC" prefix?

peculiar commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Data Sink/APCPassiveDataSink.m:
```diff
@@ -0,0 +1,477 @@
+    }
+}
+
+- (void) checkIfDataNeedsToBeFlushed
```
Method name is misleading: the data is flushed if needed and not just checking if the need is there.

Commit:
Addressing Github comments.

Commit:
Addressing PR comments.

Commit:
Addressing Github comments.

Commit:
Addressing PR comments

Commit:
Undoing changes made to reminderManager.

peculiar committed:
Merge pull request #3 from jwarmkessel/master

Addressing PR comments.

peculiar committed:
Merge branch 'apple-master' into my-master

# Conflicts:
#	APCAppCore/APCAppCore.xcodeproj/project.pbxproj
#	APCAppCore/APCAppCore/DataSubstrate/Model/APCSchedule+Bridge.m
#	APCAppCore/APCAppCore/Startup/APCAppDelegate.m
#	APCAppCore/APCAppCore/UI/Onboarding/SignUp/APCAllSetContentViewController.m
#	APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCProfileViewController.m

peculiar commented:
@jwe-apple Merge conflicts resolved.

Commit:
Addressing PR comments.

peculiar committed:
Merge pull request #4 from jwarmkessel/master

Addressing PR comments.

peculiar committed:
Removed accidentally included file.

peculiar committed:
Merge branch 'master' of https://github.com/peculiar/AppCore into my-master

# Conflicts:
#	APCAppCore/APCAppCore/Library/Objects/APCTasksReminderManager.m

peculiar committed:
Revert "Removed accidentally included file."

This reverts commit ae4a4bc043cd2e424abc01b0a8a04fac133c0f24.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/HKWorkout+APCHelper.m:
```diff
@@ -0,0 +1,111 @@
+        @(14)   : @"HKWorkoutActivityTypeDance",
+        @(15)   : @"HKWorkoutActivityTypeDanceInspiredTraining", // Pilates, Barre, Feldenkrais, etc.
+        @(16)   : @"HKWorkoutActivityTypeElliptical",
+        @(17)   : @"HKWorkoutActivityTypeEquestrianSports", // Polo, Horse Racing, Horse Riding, etc.
```
Hm, ok. Would be better to use a switch without a default: case for generating these mappings, so that if new workout activity type is added, you catch it. Could also use a macro so that the enum name is guaranteed to exactly match the string.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCCoreMotionBackgroundDataCollector.m:
```diff
@@ -0,0 +1,153 @@
+*/
+- (void)start
+{
+    if (!self.motionActivityManager)
```
Hm. This method is called on a background thread at background QOS, according to the APCPassiveDataCollector. Yet this check is not threadsafe; in fact the entire state of this collector is not thread-safe. What is to stop multiple calls to -startCollecting on the passive data collector colliding?

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NewDataCollector/Collectors/APCDisplacementTrackingCollector.m:
```diff
@@ -0,0 +1,175 @@
+        {
+            //  The user has denied use of location
+            [manager stopUpdatingLocation];
+            //  The app delegate should be notified so that other components (eg, Profile) could
```
Wait - this code doesn't actually do anything to notify the app delegate or anyone else. Where does the error go?

jwe-apple commented:
Merging despite the above comments so we can proceed elsewhere.
